napari/layers/shapes/shapes.py

```diff
@@ -3133,3 +3134,10 @@ def to_labels(self, labels_shape=None):
         labels = self._data_view.to_labels(labels_shape=labels_shape)
 
         return labels
+
+
+def warmup_numba_cache():
+    from napari.layers.shapes._shapes_utils import acc_generate_2D_edge_meshes
+
+    if acc_generate_2D_edge_meshes is not None:
+        acc_generate_2D_edge_meshes(np.array([[0, 0], [1, 1], [0, 1]]))
```

Comment:
Have you tested this in a new env @Czaki? I suspect this might not help because:

```python
In [1]: np.array([[0, 0], [1, 1], [0, 1]]).dtype
Out[1]: dtype('int64')
```

and I believe numba does not generalise across dtypes (but I am not sure). So I would instead propose:

```suggestion
        for dtype in [np.float32, np.float64]:
            acc_generate_2D_edge_meshes(np.array([[0, 0], [1, 1], [0, 1]]).astype(dtype))
```

I don't believe we use float16 or integer dtypes so that should cover all the use cases...

Reply:
It removes painting lag, but I do not tested on fresh env.
It should be only float32.

Reply:
I made a fresh env and installed this PR as is -- no change from JNI -- and there is a marked delay when making a shapes layer in the gui *and* there is the issue with first shape painting.
(note i have triangle installed and partseg backend)

Reply:
Is there an easy way to delete the numba cache from an existing env to make testing easier?

Reply:
Because the numba cache can live in a bunch of places, it doesn't seem trivial to make it foolproof:

https://stackoverflow.com/questions/44131691/how-to-clear-cache-or-force-recompilation-in-numba

In my testing I found the same as you @psobolewskiPhD. I'm confident enough in my solution that I don't expect I'll need to recreate the env more than once, so I'm just gonna go with that for now and report back, and push if it works.

Reply:
> I'm confident enough in my solution that I don't expect I'll need to recreate the env more than once

I had to open my big mouth didn't I 😂 

I added `.astype(np.float32)` to the `np.array()` call and still observed *both* pauses. I'm gonna check now the remaining code paths because I suspect we are not exercising the full set of numba compiled functions with just this one call.

Ideally, this would happen in a new thread, too, to avoid the beachball at layer-creation time. But I think for that we need to start managing a module- or class-level threadpool which is a longer job. I do think in the meantime that it is preferable to have the beachball at layer creation time than at first-draw time. @psobolewskiPhD do you agree? (Assuming we get to the bottom of this and indeed eliminate the first-draw lag.)

Reply:
Here's a summary (there's lots more versions, probably corresponding to different envs) of the numba cache files (.nbi + .nbc) in my `napari/layers/shapes/__pycache__` folder:

```
_accelerated_triangulate._cross_z-18.py311.nbi
_accelerated_triangulate._cross_z-18.py311.1.nbc
_accelerated_triangulate._cross_z-18.py311.2.nbc
_accelerated_triangulate.create_box_from_bounding-604.py311.nbi
_accelerated_triangulate.create_box_from_bounding-604.py311.1.nbc
_accelerated_triangulate.generate_2D_edge_meshes-481.py311.nbi
_accelerated_triangulate.generate_2D_edge_meshes-481.py311.1.nbc
_accelerated_triangulate.remove_path_duplicates-568.py311.nbi
_accelerated_triangulate.remove_path_duplicates-568.py311.1.nbc
```

it looks like the .nbi files store information about the *existence* of a pre-compiled function, and .nbc files actually store the binary-compiled functions. The .1/.2 in the .nbc names would correspond to different dtypes. (This is just me doing archeology, I'm sure there's numba docs somewhere that would answer this for sure.)

I'm writing this down to document my process — my hypothesis is that when I make the layer, I will see some new cached files but not all of them, and when I try my first draw, the remaining ones will appear.

Reply:
Yup — actually a bit more complicated:

- After launch:

```
~/projects/napari on prefetch_numba_code(✗) at 15:53:55
$ ls /Users/jni/micromamba/envs/numba-warmup/lib/python3.12/site-packages/napari/layers/shapes/__pycache__
__init__.cpython-312.pyc                 _shapes_constants.cpython-312.pyc        points_mgui.cpython-312.pyc
_accelerated_triangulate.cpython-312.pyc _shapes_key_bindings.cpython-312.pyc     shapes.cpython-312.pyc
_mesh.cpython-312.pyc                    _shapes_mouse_bindings.cpython-312.pyc
_shape_list.cpython-312.pyc              _shapes_utils.cpython-312.pyc
```

- After creating shapes layer:

```
(numba-warmup)
~/projects/napari on prefetch_numba_code(✗) at 15:54:15
$ ls /Users/jni/micromamba/envs/numba-warmup/lib/python3.12/site-packages/napari/layers/shapes/__pycache__
__init__.cpython-312.pyc                                         _shapes_constants.cpython-312.pyc
_accelerated_triangulate.cpython-312.pyc                         _shapes_key_bindings.cpython-312.pyc
_accelerated_triangulate.generate_2D_edge_meshes-465.py312.1.nbc _shapes_mouse_bindings.cpython-312.pyc
_accelerated_triangulate.generate_2D_edge_meshes-465.py312.nbi   _shapes_utils.cpython-312.pyc
_mesh.cpython-312.pyc                                            points_mgui.cpython-312.pyc
_shape_list.cpython-312.pyc                                      shapes.cpython-312.pyc
```

- after drawing:

```
(numba-warmup)
~/projects/napari on prefetch_numba_code(✗) at 15:54:28
$ ls /Users/jni/micromamba/envs/numba-warmup/lib/python3.12/site-packages/napari/layers/shapes/__pycache__
__init__.cpython-312.pyc                                          _accelerated_triangulate.remove_path_duplicates-556.py312.nbi
_accelerated_triangulate.cpython-312.pyc                          _mesh.cpython-312.pyc
_accelerated_triangulate.create_box_from_bounding-604.py312.1.nbc _shape_list.cpython-312.pyc
_accelerated_triangulate.create_box_from_bounding-604.py312.nbi   _shapes_constants.cpython-312.pyc
_accelerated_triangulate.generate_2D_edge_meshes-465.py312.1.nbc  _shapes_key_bindings.cpython-312.pyc
_accelerated_triangulate.generate_2D_edge_meshes-465.py312.2.nbc  _shapes_mouse_bindings.cpython-312.pyc
_accelerated_triangulate.generate_2D_edge_meshes-465.py312.3.nbc  _shapes_utils.cpython-312.pyc
_accelerated_triangulate.generate_2D_edge_meshes-465.py312.nbi    points_mgui.cpython-312.pyc
_accelerated_triangulate.remove_path_duplicates-556.py312.1.nbc   shapes.cpython-312.pyc
```

So it looks like we need to figure out the .1/.2/.3 things because more than one is being generated after drawing. (These results include the .astype(np.float32) conversion before the first run:

```diff
diff --git a/napari/layers/shapes/shapes.py b/napari/layers/shapes/shapes.py
index 75033980a..7f775363c 100644
--- a/napari/layers/shapes/shapes.py
+++ b/napari/layers/shapes/shapes.py
@@ -3140,4 +3140,6 @@ def warmup_numba_cache():
     from napari.layers.shapes._shapes_utils import acc_generate_2D_edge_meshes

     if acc_generate_2D_edge_meshes is not None:
-        acc_generate_2D_edge_meshes(np.array([[0, 0], [1, 1], [0, 1]]))
+        acc_generate_2D_edge_meshes(
+            np.array([[0, 0], [1, 1], [0, 1]]).astype(np.float32)
+        )
```
)

Reply:
> it looks like the .nbi files store information about the existence of a pre-compiled function, and .nbc files actually store the binary-compiled functions. The .1/.2 in the .nbc names would correspond to different dtypes. 

Yup. From [the docs](https://numba.pydata.org/numba-doc/0.48.0/developer/caching.html#the-implementation):

> The index of the cache is stored in a `.nbi` file, with one index per function, and it lists all the overloaded signatures compiled for the function. The object code is stored in files with an `.nbc` extension, one file per overload.

This would suggest that we actually pass multiple different dtypes to the function when drawing, which is bad.

Reply:
After the #7470, the only datatype needed is float64.

Reply:
@Czaki I know that is the *intent* of #7470, but I have just tested this again with 4b73cb0 and

- I still see the lag on first-draw, in addition to the lag on shapes layer creation
- I see one cached version of generate_2D_edge_meshes-465.py312 after layer creation
- I see many compiled versions of generate_2D_edge_meshes-465.py312 after drawing
- in addition to the other compiled functions which we should also pre-compile

Please read [this message](https://github.com/napari/napari/pull/7541#discussion_r1923072187) carefully because it shows that we are missing something significant, and it is still happening:

```
~/projects/napari on prefetch_numba_code(✗) at 23:46:36
$ ls /Users/jni/micromamba/envs/numba-warmup/lib/python3.12/site-packages/napari/layers/shapes/__pycache__
__init__.cpython-312.pyc
_accelerated_triangulate.cpython-312.pyc
_mesh.cpython-312.pyc
_shape_list.cpython-312.pyc
_shapes_constants.cpython-312.pyc
_shapes_key_bindings.cpython-312.pyc
_shapes_mouse_bindings.cpython-312.pyc
_shapes_utils.cpython-312.pyc
points_mgui.cpython-312.pyc
shapes.cpython-312.pyc
(numba-warmup)
~/projects/napari on prefetch_numba_code(✗) at 23:48:30
$ ls /Users/jni/micromamba/envs/numba-warmup/lib/python3.12/site-packages/napari/layers/shapes/__pycache__
__init__.cpython-312.pyc
_accelerated_triangulate.cpython-312.pyc
_accelerated_triangulate.generate_2D_edge_meshes-465.py312.1.nbc
_accelerated_triangulate.generate_2D_edge_meshes-465.py312.nbi
_mesh.cpython-312.pyc
_shape_list.cpython-312.pyc
_shapes_constants.cpython-312.pyc
_shapes_key_bindings.cpython-312.pyc
_shapes_mouse_bindings.cpython-312.pyc
_shapes_utils.cpython-312.pyc
points_mgui.cpython-312.pyc
shapes.cpython-312.pyc
(numba-warmup)
~/projects/napari on prefetch_numba_code(✗) at 23:48:43
$ ls /Users/jni/micromamba/envs/numba-warmup/lib/python3.12/site-packages/napari/layers/shapes/__pycache__
__init__.cpython-312.pyc
_accelerated_triangulate.cpython-312.pyc
_accelerated_triangulate.create_box_from_bounding-604.py312.1.nbc
_accelerated_triangulate.create_box_from_bounding-604.py312.nbi
_accelerated_triangulate.generate_2D_edge_meshes-465.py312.1.nbc
_accelerated_triangulate.generate_2D_edge_meshes-465.py312.2.nbc
_accelerated_triangulate.generate_2D_edge_meshes-465.py312.3.nbc
_accelerated_triangulate.generate_2D_edge_meshes-465.py312.nbi
_accelerated_triangulate.remove_path_duplicates-556.py312.1.nbc
_accelerated_triangulate.remove_path_duplicates-556.py312.nbi
_mesh.cpython-312.pyc
_shape_list.cpython-312.pyc
_shapes_constants.cpython-312.pyc
_shapes_key_bindings.cpython-312.pyc
_shapes_mouse_bindings.cpython-312.pyc
_shapes_utils.cpython-312.pyc
points_mgui.cpython-312.pyc
shapes.cpython-312.pyc
```

Reply:
Against conjunction of two things... 
1. Lack of src layout 
2. MacOS behave worse than linux

Reply:
> MacOS behave worse than linux

Which bit is macOS specific?

Reply:
On my macOS, the observed lag was much bigger than on my Linux. It looks like compilation takes longer here.